src/uuid/uuidbuilder.rs

```diff
@@ -33,174 +32,196 @@ pub struct UUIDBuilder {
-            .to_hyphenated_string()
-            .starts_with("018c6844-68f8-8000-"));
-
-        let uuid_for_same_instant = builder.build_with_instant(instant);
-        assert!(uuid_for_same_instant.is_uprotocol_uuid());
-        // same instant, version (8) and _incremented_ counter (001) should show up in UUID
-        assert!(uuid_for_same_instant
-            .to_hyphenated_string()
-            .starts_with("018c6844-68f8-8001-"));
-    }
+    use async_std::task;
+    use std::collections::HashSet;
+    use std::time::Instant;
 
     #[test]
     fn test_uuid_for_constant_random() {
         let factory = UUIDBuilder::new();
-        let uuid1 = factory.build();
-        let uuid2 = factory.build();
+        let uuid1 = factory.build_from_instance();
+        let uuid2 = factory.build_from_instance();
         assert_eq!(uuid1.lsb, uuid2.lsb);
     }
 
-    #[test]
-    #[should_panic]
-    fn test_uuid_panics_for_invalid_timestamp() {
-        // maximum value that can be stored in a 48-bit timestamp (in milliseconds)
-        let max_48_bit_unix_ts_ms = (1u64 << 48) - 1;
+    #[async_std::test]
+    async fn test_uuidbuilder_concurrency_safety_with_lsb_check() {
+        // we get near to but do not cross over 4096 messages / ms
+        let num_tasks = 10; // Number of concurrent tasks
+        let uuids_per_task = 409; // Adjusted to ensure total does not exceed 4095 in a ms burst (10 * 409 = 4090 max UUIDs per ms)
+
+        // Obtain a UUID before spawning tasks to determine the expected LSB
+        let expected_lsb = UUIDBuilder::build().lsb;
+
+        let mut tasks = Vec::new();
+
+        // Record the start time before any tasks are spawned
+        let start = Instant::now();
+
+        for task_id in 0..num_tasks {
+            let expected_lsb_clone = expected_lsb;
+            let task = task::spawn(async move {
+                let mut local_uuids = Vec::new();
+                for _ in 0..uuids_per_task {
+                    let uuid = UUIDBuilder::build();
+                    assert_eq!(
+                        uuid.lsb, expected_lsb_clone,
+                        "LSB does not match the expected value."
+                    );
+                    local_uuids.push((task_id, uuid));
+                }
+                local_uuids
+            });
+            tasks.push(task);
+        }
+
+        // Await all tasks and collect their results
+        let results = futures::future::join_all(tasks).await;
+
+        // Record the end time after all tasks are completed
+        let end = Instant::now();
 
-        // add 1 millisecond to the maximum duration, to overflow
-        let overflowed_48_bit_unix_ts_ms = max_48_bit_unix_ts_ms + 1;
+        let duration = end.duration_since(start);
 
-        let builder = UUIDBuilder::new();
-        let _uprotocol_uuid_past_max_unix_ts_ms =
-            builder.build_with_instant(overflowed_48_bit_unix_ts_ms);
+        println!("All tasks completed in {:?}.", duration);
```

Comment:
I prefer to use `log` instead of `println`

Reply:
Yup, should swap in `log` :+1: